### PR TITLE
[SPARK-27299][GRAPH][WIP] Spark Graph API design proposal

### PR DESCRIPTION
**Pull request by s1ck**
## What changes were proposed in this pull request?

This PR demonstrates a prototypical implementation of the new [Spark Graph API](https://issues.apache.org/jira/browse/SPARK-25994). The PR should mainly be used to discuss the API proposed in this [GoogleDoc](https://docs.google.com/document/d/1Wxzghj0PvpOVu7XD1iA8uonRYhexwn18utdcTxtkxlI/edit#heading=h.fly32gjytkss). This PR is not intended to be merged.

The PR introduces two modules:

* spark-graph-api (containing the API to be discussed)
* spark-cypher (a prototypical implementation of spark-graph-api)

Please use the PR and/or the GoogleDoc to comment the content of spark-graph-api. There will be follow-up PRs for spark-cypher.

## How was this patch tested?

spark-cypher has been tested using the [openCypher Technology Compatibility Kit](https://github.com/opencypher/openCypher/tree/master/tck)

## Contributors

Design, documentation and implementation have been a collaborative effort:

Co-Authored-By: Xiangrui Meng <meng@databricks.com>
Co-Authored-By: Max Kießling <max.kiessling@neotechnology.com>
Co-Authored-By: Mats Rydberg <mats@neotechnology.com>
Co-Authored-By: Philip Stutz <philip.stutz@gmail.com>
Co-Authored-By: Sören Reichardt <soren.reichardt@neotechnology.com>
Co-Authored-By: Jonatan Jäderberg <jonatan.jaderberg@gmail.com>
Co-Authored-By: Tobias Johansson <tobias.johansson@neotechnology.com>
Co-Authored-By: Alastair Green <alastair.green@neo4j.com>

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Although this is WIP, could you add LICENSE?
```
========================================================================
Running Apache RAT checks
========================================================================
Attempting to fetch rat
Could not find Apache license headers in the following files:
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/api/pom.xml
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/api/src/main/scala/org/apache/spark/graph/api/CypherResult.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/api/src/main/scala/org/apache/spark/graph/api/CypherSession.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/api/src/main/scala/org/apache/spark/graph/api/GraphElementFrame.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/api/src/main/scala/org/apache/spark/graph/api/PropertyGraph.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/api/src/main/scala/org/apache/spark/graph/api/PropertyGraphType.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/pom.xml
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/SparkCypherEntity.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/SparkCypherFunctions.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/SparkCypherRecords.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/SparkCypherResult.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/SparkCypherSession.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/SparkEntityTable.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/SparkGraphDirectoryStructure.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/SparkTable.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/adapters/MappingAdapter.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/adapters/RelationalGraphAdapter.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/adapters/SchemaAdapter.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/conversions/CypherValueEncoders.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/conversions/ExprConversions.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/conversions/RowConversion.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/conversions/StringEncodingUtilities.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/conversions/TemporalConversions.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/conversions/TypeConversions.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/io/ReadWriteGraph.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/udfs/TemporalUdfs.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/main/scala/org/apache/spark/cypher/util/HadoopFSUtils.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/test/resources/tck/failing_blacklist
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/test/resources/tck/failure_reporting_blacklist
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/test/resources/tck/temporal_blacklist
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/test/resources/tck/wont_fix_blacklist
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/test/scala/org/apache/spark/cypher/GraphExamplesSuite.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/test/scala/org/apache/spark/cypher/PropertyGraphReadWrite.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/test/scala/org/apache/spark/cypher/SharedCypherContext.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/test/scala/org/apache/spark/cypher/construction/ScanGraphFactory.scala
 !????? /home/jenkins/workspace/SparkPullRequestBuilder@2/graph/cypher/src/test/scala/org/apache/spark/cypher/tck/SparkCypherTckSuite.scala
[error] running /home/jenkins/workspace/SparkPullRequestBuilder@2/dev/check-license ; received return code 1
```

**Comment by felixcheung** [comment]
so this is adding a `graph` directory next to `graphx`?

**Comment by s1ck** [comment]
@felixcheung The [SPIP](https://issues.apache.org/jira/browse/SPARK-25994) involves Cypher querying and graph algorithms. The `graphx` directory will eventually be removed.

**Comment by dongjoon-hyun** [comment]
@s1ck . How do you build this PR in your environment? Jenkins seems to complain with `Cyclic reference` issue.
```scala
[error] Cyclic reference involving 
[error]    Project(id cypher, base: /home/jenkins/workspace/SparkPullRequestBuilder/graph/cypher, dependencies: List(ResolvedClasspathDependency(ProjectRef(file:/home/jenkins/workspace/SparkPullRequestBuilder/,core),None), 
```

**Comment by s1ck** [comment]
@dongjoon-hyun I'm using Maven/Intellij to build and run. I used the same `sbt.project.name` for the two modules .. I guess this lead to the cyclic reference error.

**Comment by s1ck** [comment]
@dongjoon-hyun The cyclic reference is fixed. I won't dive into fixing Scala style tests for now since this is WIP and open for discussion. Btw, do you know if there is a Spark code style xml for IntelliJ?

**Comment by dongjoon-hyun** [comment]
IntelliJ already understands `checkstyle.xml`. If you open the project with IntelliJ in Apache Spark project, you must see the red line warnings.
- https://github.com/apache/spark/blob/master/dev/checkstyle.xml

The current warnings looks like two types mostly. You can fix them easily.
- Max line length (100).
- Import ordering and grouping (`java` -> `scala` -> `3rd party` -> `org.apache.spark`)

**Comment by dongjoon-hyun** [comment]
Hi, @mengxr . Could you help this PR to pass the Jenkins, please?

**Comment by mengxr** [comment]
@dongjoon-hyun This PR is a prototype for API and design discussions. We should break it down into smaller ones after we reach an agreement on the API and design. I don't think we can merge this one directly.

**Comment by dongjoon-hyun** [comment]
Thank you, @mengxr . I see.

File: graph/cypher/src/main/scala/org/apache/spark/cypher/SparkCypherRecords.scala [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  def df: DataFrame = table.df
+
+  override def cache(): SparkCypherRecords = {
+    df.cache()
+    this
+  }
+
+  override def toString: String = {
+    if (header.isEmpty) {
+      s"CAPSRecords.empty"
```
Comment (kiszk) [inline graph/cypher/src/main/scala/org/apache/spark/cypher/SparkCypherRecords.scala]: nit: `s` is not necessary

File: graph/cypher/src/main/scala/org/apache/spark/cypher/SparkGraphDirectoryStructure.scala [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  def nodeTableDirectoryName(labels: Set[String]): String = concatDirectoryNames(labels.toSeq.sorted)
+
+  def relKeyTableDirectoryName(relKey: String): String = relKey.encodeSpecialCharacters
+
+  def concatDirectoryNames(seq: Seq[String]): String = {
+    if (seq.isEmpty) {
+      noLabelNodeDirectoryName
+    } else {
+      // TODO: Find more elegant solution for encoding underline characters
+      seq.map(_.replace("_", "--UNDERLINE--")).mkString("_").encodeSpecialCharacters
```
Comment (kiszk) [inline graph/cypher/src/main/scala/org/apache/spark/cypher/SparkGraphDirectoryStructure.scala]: nit: Is it better to define a variable to hold `"--UNDERLINE--` and use it at all of the references for consistency?

File: graph/cypher/src/main/scala/org/apache/spark/cypher/adapters/MappingAdapter.scala [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    def toNodeMapping: ElementMapping = NodeMappingBuilder
+      .on(nodeDf.idColumn)
+      .withImpliedLabels(nodeDf.labelSet.toSeq: _*)
+      .withPropertyKeyMappings(nodeDf.properties.toSeq:_*)
+      .build
+  }
+
+  implicit class RichRelationshipDataFrame(val relDf: RelationshipFrame) extends AnyVal {
+    def toRelationshipMapping: ElementMapping = RelationshipMappingBuilder
+        .on(relDf.idColumn)
```
Comment (kiszk) [inline graph/cypher/src/main/scala/org/apache/spark/cypher/adapters/MappingAdapter.scala]: 2-indent?

File: graph/cypher/src/main/scala/org/apache/spark/cypher/conversions/ExprConversions.scala [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+      * argument.
+      *
+      * Iff the expression has `expr.nullInNullOut == true`, then any child being mapped to `null` will also result in
+      * the parent expression being mapped to null.
+      *
+      * For these expressions the `withConvertedChildren` function is guaranteed to not receive any `null`
+      * values from the evaluated children.
+      */
+    def nullSafeConversion(expr: Expr)(withConvertedChildren: Seq[Column] => Column)
+      (implicit header: RecordHeader, df: DataFrame, parameters: CypherMap): Column = {
```
Comment (kiszk) [inline graph/cypher/src/main/scala/org/apache/spark/cypher/conversions/ExprConversions.scala]: 4-indent?

**Comment by s1ck** [comment]
Hey Kazuaki. Very nice of you to find the time to look over the PR. However, this PR is not intended to be merged. It's main purpose is to have a running PoC while we implement the graph features in smaller PRs. The latest one is introducing the API for property graph construction (https://github.com/apache/spark/pull/24851). Please feel free to have a look. I'll address your comments when we start working on [SPARK-27309](https://issues.apache.org/jira/browse/SPARK-27309). @kiszk

**Pull request closed without merging**